### PR TITLE
Compressed source fixes and tests

### PR DESCRIPTION
**Pull request by dhalperi**
This is a backport of

https://github.com/apache/incubator-beam/pull/794

and

https://github.com/apache/incubator-beam/pull/504

### CODE REVIEW

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java [hunk 1]
```diff
@@ -522,6 +522,31 @@ public void testGZIPReadWhenUncompressed() throws Exception {
   }
 
   @Test
+  public void testGZIPReadWhenCompressed() throws Exception {
```
Comment (bjchambers) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: Can we add a similar unit test for BZIP2?

**Comment by bjchambers** [comment]
Some suggestions for more testing, but otherwise looks good.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java [hunk 2]
```diff
@@ -522,6 +522,31 @@ public void testGZIPReadWhenUncompressed() throws Exception {
...
+    final int numLines = 1000;
```
Comment (kennknowles) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: Comment about why 1000 so no one tries to reduce it.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java [hunk 3]
```diff
@@ -522,6 +522,31 @@ public void testGZIPReadWhenUncompressed() throws Exception {
...
+    List<String> expected = new ArrayList<>();
+    try (PrintStream writer = new PrintStream(
+        new GZIPOutputStream(new FileOutputStream(tmpFile)))) {
+      for (int i = 0; i < numLines; ++i) {
+        String line = "word" + i;
+        writer.println(line);
+        expected.add(line);
+      }
+    }
+
```
Comment (kennknowles) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: Hereabouts assert that the compressed size is less than the uncompressed size.

**Comment by kennknowles** [comment]
LGTM with some testing comments. The only one I'd block merge on is the assertion that the compressed size is smaller than uncompressed size.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java (hunk 1 above)
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: done

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java [hunk 4]
```diff
@@ -522,6 +522,31 @@ public void testGZIPReadWhenUncompressed() throws Exception {
...
+    Pipeline p = TestPipeline.create();
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.GZIP);
```
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: done

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java (hunk 2 above)
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: done

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java (hunk 3 above)
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: done

**Comment by dhalperi** [comment]
PTAL -- reworked the test entirely in 5183e79

**Comment by kennknowles** [comment]
LGTM. Some of the test methods could be broken up for a more granular view in IDEs, Jenkins, etc, but let's get it merged instead of splitting hairs.

**Pull request merged**